src/accelerate/checkpointing.py

```diff
@@ -214,3 +230,322 @@ def load_custom_state(obj, path, index: int = 0):
+        if last_block_size + weight_size > max_shard_size:
+            sharded_state_dicts.append({})
+            last_block_size = 0
+
+        sharded_state_dicts[-1][key] = weight
+        last_block_size += weight_size
+        total_size += weight_size
+        storage_id_to_block[storage_id] = len(sharded_state_dicts) - 1
+
+    # If we only have one shard, we return it
+    if len(sharded_state_dicts) == 1:
+        return {weights_name: sharded_state_dicts[0]}, None
+
+    # Otherwise, let's build the index
+    weight_map = {}
+    shards = {}
+    for idx, shard in enumerate(sharded_state_dicts):
+        shard_file = weights_name.replace(".bin", f"-{idx+1:05d}-of-{len(sharded_state_dicts):05d}.bin")
+        shard_file = shard_file.replace(
+            ".safetensors", f"-{idx + 1:05d}-of-{len(sharded_state_dicts):05d}.safetensors"
+        )
+        shards[shard_file] = shard
+        for key in shard.keys():
+            weight_map[key] = shard_file
+
+    # Add the metadata
+    metadata = {"total_size": total_size}
+    index = {"metadata": metadata, "weight_map": weight_map}
+    return shards, index
+
+
+def unwrap_model(model: nn.Module) -> nn.Module:
+    """
+    Recursively unwraps a model from potential containers (as used in distributed training).
+
+    Args:
+        model (`torch.nn.Module`): The model to unwrap.
+    """
+    # since there could be multiple levels of wrapping, unwrap recursively
+    if hasattr(model, "module"):
+        return unwrap_model(model.module)
+    else:
+        return model
+
+
+def save_model(
+    model: nn.Module,
+    save_directory: Union[str, os.PathLike],
+    is_main_process: bool = True,
+    state_dict: Optional[dict] = None,
+    max_shard_size: Union[int, str] = "10GB",
+    save_function: Callable = torch.save,
+    safe_serialization: bool = False,
+    variant: Optional[str] = None,
+    keys_to_ignore_on_save: Optional[List[str]] = None,
+):
+    """
+    Save a model and its configuration file to a directory, so that it can be re-loaded using load_checkpoint_in_model
+
+    Arguments:
+        save_directory (`str` or `os.PathLike`):
+            Directory to which to save. Will be created if it doesn't exist.
+        is_main_process (`bool`, *optional*, defaults to `True`):
+            Whether the process calling this is the main process or not. Useful when in distributed training like TPUs
+            and need to call this function on all processes. In this case, set `is_main_process=True` only on the main
+            process to avoid race conditions.
+        state_dict (nested dictionary of `torch.Tensor`):
+            The state dictionary of the model to save. Will default to `self.state_dict()`, but can be used to only
+            save parts of the model or if special precautions need to be taken when recovering the state dictionary of
+            a model (like when using model parallelism).
```

Comment:
I don't think we should support this argument here. This is specific to some things in the Trainer.

Reply:
Fixed [8d853b4](https://github.com/huggingface/accelerate/pull/1641/commits/8d853b4c7918360b91877737c363b5722b4c8e48)